app/mailers/application_mailer.rb

```diff
@@ -1,6 +1,16 @@
 class ApplicationMailer < Mail::Notify::Mailer
   GENERIC_NOTIFY_TEMPLATE = '2744ea53-34f1-431f-8173-8388fadd826a'.freeze
 
+  rescue_from Notifications::Client::RequestError do
+    # WARNING: this needs to be a block, otherwise the exception won't be
+    # re-raised and we won't be notified via Sentry, and the job won't retry.
+    #
+    # @see https://github.com/rails/rails/issues/39018
+    email = Email.find(self.headers['email-log-id'].to_s)
```

Comment:
Say there's an email with an unstable `notify_ref` and notify is down. Each time we try to deliver it we generate a new ref. Each time it fails, the log-id is plucked out of the header, and we mark that attempt as `notify_error`. Once notify gets the email then it calls us back with the last `notify_ref`, we do `Email.find_by(notify_reference: blah)` and we mark that `delivered`. All the failed ones are marked `failed` and the successful one is marked `successful`. This is good.

But if we have a stable `notify_ref` eg `production_reference-request_104`, then the `find_by(notify_reference: production_reference_request_104)` _could_ find one of the failures instead of the pending one. The email log would then contain a pending email as well as some `notify_error`s and one `successful`.

Should we change `find_by(notify_reference: blah)` to `find_by(notify_reference: blah, status: 'pending')`?

Reply:
Alternatively, we could always append random hashes to `notify_ref`s

Reply:
Good catch! PR here: https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/pull/1928